### ISSUE TITLE
Why can I not add to this?

### ISSUE THREAD

**Issue by Undefined-Variables** [pH7Software/pH7-Social-Dating-CMS#801]
I need to add to this in...

     _protected/framework/Mvc/Controller/Controller.class.php

````

private function setUserNotifications()
    {
        $aNotificationCounter = [
            'count_unread_mail' => MailCoreModel::countUnreadMsg($this->session->get('member_id')),
            'count_pen_friend_request' => FriendCoreModel::getPending($this->session->get('member_id'))
        ];
        $this->view->assigns($aNotificationCounter);

        unset($aNotificationCounter);
    }
 
````

I need to add like this but it breaks the site and I can't see why...

````
private function setUserNotifications()
    {
        $aNotificationCounter = [
            'count_unread_mail' => MailCoreModel::countUnreadMsg($this->session->get('member_id')),
            'count_pen_friend_request' => FriendCoreModel::getPending($this->session->get('member_id')),
            'count_unread_gifts' => GiftsCoreModel::countUnreadGifts($this->session->get('member_id'))
        ];
        $this->view->assigns($aNotificationCounter);

        unset($aNotificationCounter);
    }

````

Am I missing something here? I can't see why this breaks the site. GiftsCoreModel exists and has this in it....

````

<?php
/**
 * @author         Pierre-Henry Soria <ph7software@gmail.com>
 * @copyright      (c) 2012-2019, Pierre-Henry Soria. All Rights Reserved.
 * @license        GNU General Public License; See PH7.LICENSE.txt and PH7.COPYRIGHT.txt in the root directory.
 * @package        PH7 / App / System / Core / Model
 */

namespace PH7;

use PH7\Framework\Mvc\Model\Engine\Db;
use PH7\Framework\Mvc\Model\Engine\Model;

// Abstract Class
class GiftsCoreModel extends Model
{
    const READ_STATUS = 0;
    const UNREAD_STATUS = 1;

    /**
     * Get the number of unread gifts.
     *
     * @param int $iProfileId
     *
     * @return int
     */
    public static function countUnreadGifts($iProfileId)
    {
        $rStmt = Db::getInstance()->prepare('SELECT COUNT(status) AS unread FROM' . Db::prefix(DbTableName::GIFT_BOX) .
            'WHERE recipient = :recipient AND status = :status AND NOT FIND_IN_SET(\'recipient\', toDelete)');

        $rStmt->bindValue(':recipient', $iProfileId, \PDO::PARAM_INT);
        $rStmt->bindValue(':status', self::UNREAD_STATUS, \PDO::PARAM_INT);
        $rStmt->execute();
        $gUnread = (int)$rStmt->fetchColumn();
        Db::free($rStmt);

        return $gUnread;
    }
}


````

**Comment by polynamaude** [comment]
@Undefined-Variables Remember a while ago I said that there's some really annoying tutorials and reading that has to be done or you'll get errors you won't understand or that you'll get to limits you won't feel...
Well there's one here...

It probably has to do with the rules regarding visibility and class hierarchy... what you can extend, inherit, what's abstract, etc...

And, like you and I repeated more than once... If you'd do more than just say **it's broken** but give some error messages or screenshot then it would be a bit easier...

**Comment by Undefined-Variables** [comment]
Hey @polynamaude 

I wish there was something in the error log cos that would answer the problem. It should work as I have copied how personal messages are handled and all else works and this is the very last piece to the puzzle and as soon as I save the new line I get 500 the server can not handle this request! I have not had an issue like this in a long time. I know how this software works and I am confident in saying I could most likely write most features in to it but this makes no sense. 

I thought I may just be overlooking something or maybe I have been looking for too long so a fresh set of eyes is never a bad thing! Driving me nuts cos there is no reason I can see stopping it working!

**Comment by polynamaude** [comment]
@Undefined-Variables Start by changer the verbose level in your PHP configuration.
This will give you more information. If you get a error 500 then it will be shown why at some point.
Run the software on a local instance with a PHP debugger (xDebug or similar).

**Comment by Undefined-Variables** [comment]
Okay now that is a good idea and gives me something to work with! What debugger would you recommend?

**Comment by Undefined-Variables** [comment]
Don't normally use any debugger tbh I still use Komodo IDE :)

**Comment by polynamaude** [comment]
@Undefined-Variables Maybe it could be time that you upgrade your environment for development. There's a pretty good reason people use development environment and not simple code editor. That's for debugging, bug finding and much more...Following issues, looking up documentation, etc...
You can use xDebug, it's free and most of the time, it gets installed with PHP as a module. You just have to activate it.

**Comment by Undefined-Variables** [comment]
@polynamaude I am happy to try other IDE's but I cant seem to find one that I like, the shortcut buttons are never correct like PHPStorm if you hold CTRL and use mouse wheel it doesn't zoom in/out you have to go in to menu and mess about, you can't save a single file its "save all" which is not always what I want to do... Eclipse I don't like the layout and VS Code is just hard work as you have to type cmds for everything... A text editor takes a long time to get comfortable with. Oh and Atom is a pile of rubbish!

**Comment by Undefined-Variables** [comment]
@polynamaude I like VS Code when working local its so fast!

**Comment by Undefined-Variables** [comment]
@polynamaude I absolutely understand where you are coming from with use of a debugger and am falling in love with VS Code its effortless :) Thanks for advice!

**Comment by pH-7** [comment]
> I need to add to this in...
> 
> ```
>  _protected/framework/Mvc/Controller/Controller.class.php
> ```
> 
> ```
> 
> private function setUserNotifications()
>     {
>         $aNotificationCounter = [
>             'count_unread_mail' => MailCoreModel::countUnreadMsg($this->session->get('member_id')),
>             'count_pen_friend_request' => FriendCoreModel::getPending($this->session->get('member_id'))
>         ];
>         $this->view->assigns($aNotificationCounter);
> 
>         unset($aNotificationCounter);
>     }
>  
> ```
> 
> I need to add like this but it breaks the site and I can't see why...
> 
> ```
> private function setUserNotifications()
>     {
>         $aNotificationCounter = [
>             'count_unread_mail' => MailCoreModel::countUnreadMsg($this->session->get('member_id')),
>             'count_pen_friend_request' => FriendCoreModel::getPending($this->session->get('member_id')),
>             'count_unread_gifts' => GiftsCoreModel::countUnreadGifts($this->session->get('member_id'))
>         ];
>         $this->view->assigns($aNotificationCounter);
> 
>         unset($aNotificationCounter);
>     }
> ```
> 
> Am I missing something here? I can't see why this breaks the site. GiftsCoreModel exists and has this in it....
> 
> ```
> 
> <?php
> /**
>  * @author         Pierre-Henry Soria <ph7software@gmail.com>
>  * @copyright      (c) 2012-2019, Pierre-Henry Soria. All Rights Reserved.
>  * @license        GNU General Public License; See PH7.LICENSE.txt and PH7.COPYRIGHT.txt in the root directory.
>  * @package        PH7 / App / System / Core / Model
>  */
> 
> namespace PH7;
> 
> use PH7\Framework\Mvc\Model\Engine\Db;
> use PH7\Framework\Mvc\Model\Engine\Model;
> 
> // Abstract Class
> class GiftsCoreModel extends Model
> {
>     const READ_STATUS = 0;
>     const UNREAD_STATUS = 1;
> 
>     /**
>      * Get the number of unread gifts.
>      *
>      * @param int $iProfileId
>      *
>      * @return int
>      */
>     public static function countUnreadGifts($iProfileId)
>     {
>         $rStmt = Db::getInstance()->prepare('SELECT COUNT(status) AS unread FROM' . Db::prefix(DbTableName::GIFT_BOX) .
>             'WHERE recipient = :recipient AND status = :status AND NOT FIND_IN_SET(\'recipient\', toDelete)');
> 
>         $rStmt->bindValue(':recipient', $iProfileId, \PDO::PARAM_INT);
>         $rStmt->bindValue(':status', self::UNREAD_STATUS, \PDO::PARAM_INT);
>         $rStmt->execute();
>         $gUnread = (int)$rStmt->fetchColumn();
>         Db::free($rStmt);
> 
>         return $gUnread;
>     }
> }
> ```

Hey @Undefined-Variables Did you import GiftsCoreModel's namespace at the top of `Controller.class.php` file? 
```php
se PH7\GiftsCoreModel;
```

You can see for instance how FriendCoreModel is imported in Controller.class.php class https://github.com/pH7Software/pH7-Social-Dating-CMS/blob/f63d9e194a3e1e7410ca7438b8cb66163f1437df/_protected/framework/Mvc/Controller/Controller.class.php#L34

If you don't import GiftsCoreModel namespace, PHP will guess that `GiftsCoreModel` has the same namespace as Controller.class.php, which is not the case.

**Comment by pH-7** [comment]
Alternatively, you can also do `\PH7\GiftsCoreModel::countUnreadGifts(` instead of importing the namespace at the top, but it makes the code slightly less readable (especially for long namespace paths).

**Comment by pH-7** [comment]
If you wish to learn more about namespaces in PHP, have a look to https://www.php.net/manual/en/language.namespaces.php (it's pretty well explained :)).

**Comment by Undefined-Variables** [comment]
@pH-7 you are awesome! Yes I know about name spaces and knew I had overlooked something. I found another temporary solution that worked but was not the right way to do this. I called the function in the notification.inc.tpl I wrote but could not get the current user id to be returned for the argument of that function so had to use $id then go to every controller and add   
    
            $this->view->id = $this->session->get('member_id');

this was a total hack and is working way to hard. I love the way you just knew the issue on first thought. Your an absolute legend. Need a friend with your knowledge on WhatsApp lol. Seriously though thank you!
